rustc suggests to use impl trait in closure parameter which doesn't work

### Code

```Rust
let a = |x: dyn Fn()| x();
```

### Current output

```Shell
--> src/lib.rs:14:17
   |
14 |     let a = |x: dyn Fn()| x();
   |                 ^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `(dyn Fn() + 'static)`
help: you can use `impl Trait` as the argument type
   |
14 |     let a = |x: impl Fn()| x();
   |                 ~~~~
help: function arguments must have a statically known size, borrowed types always have a known size
   |
14 |     let a = |x: &dyn Fn()| x();
   |                 +
```

### Desired output

```Shell
help: function arguments must have a statically known size, borrowed types always have a known size
   |
14 |     let a = |x: &dyn Fn()| x();
   |                 +
```

### Rationale and extra context

We start with some obviously invalid code, `rustc` makes two suggestions - pass `&dyn Trait` by reference - this works and use `impl Trait` which doesn't work.

```
error[E0562]: `impl Trait` is not allowed in closure parameters
  --> src/lib.rs:15:17
   |
15 |     let b = |x: impl Fn()| x();
   |                 ^^^^^^^^^
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods

For more information about this error, try `rustc --explain E0562`.
```

Usually following compiler suggestions you get closer to having working code, but here this leads to a dead end.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.85.0 (4d91de4e4 2025-02-17)
binary: rustc
commit-hash: 4d91de4e48198da2e33413efdcd9cd2cc0c46688
commit-date: 2025-02-17
host: x86_64-unknown-linux-gnu
release: 1.85.0
LLVM version: 19.1.7
```

### Anything else?

Nightly makes a similar suggestion, also broken.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"rperier"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -3018,12 +3018,23 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                                 [] => span_bug!(ty.span, "trait object with no traits: {ty:?}"),
                             };
                             let needs_parens = traits.len() != 1;
-                            err.span_suggestion_verbose(
-                                span,
-                                "you can use `impl Trait` as the argument type",
-                                "impl ",
-                                Applicability::MaybeIncorrect,
-                            );
+                            // Don't recommend impl Trait as a closure argument
+                            if let Some(hir_id) = hir_id
+                                && matches!(
+                                    self.tcx.parent_hir_node(hir_id),
+                                    hir::Node::Item(hir::Item {
+                                        kind: hir::ItemKind::Fn { .. },
+                                        ..
+                                    })
+                                )
+                            {
+                                err.span_suggestion_verbose(
+                                    span,
+                                    "you can use `impl Trait` as the argument type",
+                                    "impl ",
+                                    Applicability::MaybeIncorrect,
+                                );
+                            }
                             let sugg = if !needs_parens {
                                 vec![(span.shrink_to_lo(), format!("&{kw}"))]
                             } else {

--- a/tests/ui/traits/dont-suggest-impl-as-closure-arg.rs
+++ b/tests/ui/traits/dont-suggest-impl-as-closure-arg.rs
@@ -1,0 +1,5 @@
+// Suggestion to use impl trait in closure parameter is invalid, see issue 138932
+fn main() {
+    let c = |f: dyn Fn()| f();
+    //~^ ERROR: the size for values of type `(dyn Fn() + 'static)` cannot be known at compilation time
+}

--- a/tests/ui/traits/dont-suggest-impl-as-closure-arg.stderr
+++ b/tests/ui/traits/dont-suggest-impl-as-closure-arg.stderr
@@ -1,0 +1,16 @@
+error[E0277]: the size for values of type `(dyn Fn() + 'static)` cannot be known at compilation time
+  --> $DIR/dont-suggest-impl-as-closure-arg.rs:3:17
+   |
+LL |     let c = |f: dyn Fn()| f();
+   |                 ^^^^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `(dyn Fn() + 'static)`
+   = help: unsized fn params are gated as an unstable feature
+help: function arguments must have a statically known size, borrowed types always have a known size
+   |
+LL |     let c = |f: &dyn Fn()| f();
+   |                 +
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

